source/extensions/filters/network/redis_proxy/router_impl.cc

```diff
@@ -35,19 +35,47 @@ PrefixRoutes::PrefixRoutes(
   }
 }
 
-ConnPool::InstanceSharedPtr PrefixRoutes::upstreamPool(std::string& key) {
+ConnPool::InstanceSharedPtr PrefixRoutes::upstreamPool(Common::Redis::RespValue& key) {
+  std::string& key_str = key.asString();
   PrefixPtr value = nullptr;
   if (case_insensitive_) {
-    std::string copy(key);
+    std::string copy(key_str);
     to_lower_table_.toLowerCase(copy);
     value = prefix_lookup_table_.findLongestPrefix(copy.c_str());
   } else {
-    value = prefix_lookup_table_.findLongestPrefix(key.c_str());
+    value = prefix_lookup_table_.findLongestPrefix(key_str.c_str());
   }
 
   if (value != nullptr) {
     if (value->remove_prefix) {
-      key.erase(0, value->prefix.length());
+      key_str.erase(0, value->prefix.length());
+    }
+    return value->upstream;
+  }
+
+  return catch_all_upstream_;
+}
+
+ConnPool::InstanceSharedPtr
+PrefixRoutes::upstreamPool(std::vector<Common::Redis::RespValue>& keys) {
+  if (keys.empty()) {
+    return catch_all_upstream_;
+  }
+
+  PrefixPtr value = nullptr;
+  if (case_insensitive_) {
+    std::string copy(keys[0].asString());
+    to_lower_table_.toLowerCase(copy);
+    value = prefix_lookup_table_.findLongestPrefix(copy.c_str());
+  } else {
+    value = prefix_lookup_table_.findLongestPrefix(keys[0].asString().c_str());
+  }
+
+  if (value != nullptr) {
+    if (value->remove_prefix) {
+      for (auto& key : keys) {
+        key.asString().erase(0, value->prefix.length());
```

Comment:
This doesn't seem safe to assume blindly that all keys have the same prefix. If the same selection logic were applied to each key individually, and the results end up with different ConnPool::Instance pointers, then nullptr should be returned. If they all resolve to the same ConnPool::Instance, then remove_prefix logic could be individually applied to each key depending on its value in the prefix_lookup_table. Just my $0.02 worth...

Reply:
Hey Mitch! No worries, thanks for looking! The eval style commands that have multiple keys only makes sense to be run on a single upstream and routed through a single ConnPool::Instance. Other multi key commands like MGET, MSET, or SplitKeysSumResult style get fanned out to each instance and the result is joined before returning. However, for eval we send all these keys as arguments to the given script to be executed on a single instance. 

That's why we currently only look at the first key sent to eval and route the whole batch of keys to one upstream. And if remove_prefix is set, it would only remove it for the first key as well. This PR makes it so that it still routes based on the first key but it removes the prefix on all keys in the command. 

Your concern is still true that if the caller sends no prefix/some other prefix for the other keys then they will get mangled. But like I mentioned not having the keys prefixed the same makes little sense. We are using https://github.com/resque/redis-namespace to prefix our keys for all commands and I suspect other such implementations will have the same behavior where every KEY in the RESP protocol is prefixed. Perhaps this is something we need to clarify in the docs.

Reply:
I think the safety check to make sure all the keys belongs to the same upstream is more cautious, and it would make sense to return nullptr otherwise. WDYT @inf-rno?

Reply:
Yup, we can definitely do that. But I suspect this will be too cautious and do busy work for normal use cases. If we're not concerned about the performance impact of double checking the keys then I'll add the check. But we could also solve this by convention/docs. I feel like we should use the fact that all keys to a script must always route to a single upstream.